### PR TITLE
services/horizon: Add include_transactions parameter to operations endpoint

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

If include_transactions is true then transaction data will be included in the operations response

### CODE REVIEW

File: protocols/horizon/operations/main.go [hunk 1]
```diff
@@ -46,12 +47,13 @@ type Base struct {
...
-	Type                  string    `json:"type"`
-	TypeI                 int32     `json:"type_i"`
-	LedgerCloseTime       time.Time `json:"created_at"`
-	TransactionHash       string    `json:"transaction_hash"`
+	TransactionSuccessful bool                 `json:"transaction_successful"`
+	SourceAccount         string               `json:"source_account"`
+	Type                  string               `json:"type"`
+	TypeI                 int32                `json:"type_i"`
+	LedgerCloseTime       time.Time            `json:"created_at"`
+	TransactionHash       string               `json:"transaction_hash"`
```
[the remark this answers is not in the document]
Comment (poliha) [reply protocols/horizon/operations/main.go]: @bartekn I remember we added `transaction_hash` a while back because we didn't want user to make an extra call for this. IIRC this was because of the bridge server at the time.

File: services/horizon/CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@ bumps.  A breaking change will get clearly notified in this log.
 
 ## Unreleased
 
+* Add `include_transactions` parameter to operations endpoints.
```
[the remark this answers is not in the document]
Comment (poliha) [reply services/horizon/CHANGELOG.md]: I like this proposal as well, but if were were to leave it I suggest that it should be `include_transaction` given that operations can only have one transaction

File: services/horizon/internal/db2/history/operation.go [hunk 3]
```diff
@@ -64,22 +65,54 @@ func (q *Q) OperationFeeStats(currentSeq int32, dest *FeeStats) error {
...
+		query.sql = selectOperation
+	}
+
+	return query
 }
 
 // OperationByID loads a single operation with `id` into `dest`
-func (q *Q) OperationByID(dest interface{}, id int64) error {
-	sql := selectOperation.
+func (q *Q) OperationByID(includeTransactions bool, id int64) (Operation, *Transaction, error) {
```
Comment (poliha) [inline services/horizon/internal/db2/history/operation.go]: The description needs to change to match the method signature

File: protocols/horizon/operations/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartekn) [reply protocols/horizon/operations/main.go]: Yes, but now users can set request param to include transaction data in the response. I think there's no need to duplicate fields (`operation.transaction_hash`/`transaction.hash`). Also no extra call required.

**Comment by tamirms** [comment]
@poliha @bartekn could you take another look? I've pushed a commit addressing the feedback

File: services/horizon/internal/db2/history/operation.go [hunk 4]
```diff
@@ -171,55 +188,73 @@ func (q *OperationsQ) Page(page db2.PageQuery) *OperationsQ {
...
+			return nil, nil, errors.Errorf("Corrupted data! `successful=true` but returned transaction is not success: %s %s", o.TransactionHash, o.TxResult)
 		}
 
 		if !o.IsTransactionSuccessful() && resultXDR.Result.Code == xdr.TransactionResultCodeTxSuccess {
-			return errors.Errorf("Corrupted data! `successful=false` but returned transaction is success: %s %s", o.TransactionHash, o.TxResult)
+			return nil, nil, errors.Errorf("Corrupted data! `successful=false` but returned transaction is success: %s %s", o.TransactionHash, o.TxResult)
+		}
+	}
+
+	if q.includeTransactions && len(transactionIDs) > 0 {
```
Comment (poliha) [inline services/horizon/internal/db2/history/operation.go]: Is it possible to just use `TransactionByHash` here? then we wont need a new method. 
Alternatively, can we update `selectOperation` query to get the transaction fields that we want given that we already join the transactions table there? My knowledge of horizon is rusty so not entirely sure if that will be the best approach. Feel free to let me know if your current approach is better. What do you think @bartekn @tamirms ?

Comment (tamirms) [reply services/horizon/internal/db2/history/operation.go]: > Is it possible to just use TransactionByHash here?

`TransactionByHash` returns a single transaction. I could rewrite `TransactionByHash` as `TransactionsByHashes` so that it can return multiple transactions. Is that what you had in mind?

> Alternatively, can we update selectOperation query to get the transaction fields that we want given that we already join the transactions table there

that is what I did in the previous commit. do you prefer that version of the code?

File: services/horizon/internal/actions_operation.go [hunk 5]
```diff
@@ -92,13 +105,37 @@ func (action *OperationIndexAction) SSE(stream *sse.Stream) error {
...
 	action.ValidateCursorAsDefault()
 	action.AccountFilter = action.GetAddress("account_id")
 	action.LedgerFilter = action.GetInt32("ledger_id")
 	action.TransactionFilter = action.GetStringFromURLParam("tx_id")
 	action.PagingParams = action.GetPageQuery()
 	action.IncludeFailed = action.GetBool("include_failed")
+	if parsed, err := parseJoinField(&action.Action.Base); err != nil {
+		action.Err = err
+		return
+	} else {
```
[the remark this answers is not in the document]
Comment (tamirms) [reply services/horizon/internal/actions_operation.go]: so you would prefer if the code was rewritten as below?

```
parsed, err := parseJoinField(&action.Action.Base)
if err != nil {
	action.Err = err
	return
}
action.IncludeTransactions = parsed[joinTransactions]
```

File: services/horizon/internal/db2/history/operation.go [hunk 6]
```diff
@@ -64,22 +64,39 @@ func (q *Q) OperationFeeStats(currentSeq int32, dest *FeeStats) error {
 
 // Operations provides a helper to filter the operations table with pre-defined
 // filters.  See `OperationsQ` for the available filters.
-func (q *Q) Operations() *OperationsQ {
-	return &OperationsQ{
-		parent:        q,
-		sql:           selectOperation,
-		opIdCol:       "hop.id",
-		includeFailed: false,
+func (q *Q) Operations(includeTransactions bool) *OperationsQ {
```
Comment (bartekn) [inline services/horizon/internal/db2/history/operation.go]: I think we should move setting `includeTransactions` to a method. It's not clear what true/false means in the code using this method, you need to go and check the godoc or a method. In short instead of:
```go
query := q.Operations(true).
		ForAccount(accountID)
```
I think we should do:
```go
query := q.Operations().
		JoinTransations().
		ForAccount(accountID)
```
the later is more clear for the person reading the code.

File: services/horizon/internal/db2/history/operation.go (hunk 4 above)
Comment (bartekn) [reply services/horizon/internal/db2/history/operation.go]: I think we should stay with the current method (`TransactionsByIDs`). We shouldn't use `TransactionByHash` because it will send up to 200 additional queries (one for each operation loaded) increasing the execution time. Alternatively we could create `TransactionsByHashes` but `TransactionsByIDs` is exactly the same, just using a different unique ID of a transaction.

File: services/horizon/internal/db2/history/operation_test.go [hunk 7]
```diff
@@ -240,14 +234,47 @@ func TestExtraChecksOperationsTransactionSuccessfulFalseResultTrue(t *testing.T)
...
 		IncludeFailed()
 
-	err = query.Select(&operations)
+	_, _, err = query.Fetch()
 	tt.Assert.Error(err)
 	tt.Assert.Contains(err.Error(), "Corrupted data! `successful=false` but returned transaction is success")
 }
+
+// TestOperationIncludeTransactions tests that transactions are included when fetching records from the db.
+func TestOperationIncludeTransactions(t *testing.T) {
```
Comment (bartekn) [inline services/horizon/internal/db2/history/operation_test.go]: Can we add tests that check if `operations[i]` is really part of `transactions[i]` (by comparing hash or id)? Same think for `OperationByID`.

File: services/horizon/internal/resourceadapter/operations_test.go [hunk 8]
```diff
@@ -23,20 +23,20 @@ func TestPopulateOperation_Successful(t *testing.T) {
...
-	PopulateBaseOperation(ctx, &dest, row, ledger)
+	PopulateBaseOperation(ctx, &dest, row, nil, ledger)
 	assert.True(t, dest.TransactionSuccessful)
 
 	dest = operations.Base{}
 	val = false
 	row = history.Operation{TransactionSuccessful: &val}
 
-	PopulateBaseOperation(ctx, &dest, row, ledger)
+	PopulateBaseOperation(ctx, &dest, row, nil, ledger)
```
Comment (bartekn) [inline services/horizon/internal/resourceadapter/operations_test.go]: Could you add a test when `transactionRow` is not nil?

File: services/horizon/internal/actions_operation.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (poliha) [reply services/horizon/internal/actions_operation.go]: Yes. Just my preference though.. plus less lines :wink:

File: services/horizon/internal/db2/history/operation.go (hunk 4 above)
Comment (poliha) [reply services/horizon/internal/db2/history/operation.go]: Okay makes sense... I had not realised that we were returning multiple transactions. @tamirms  your way is fine

**Comment by tamirms** [comment]
@poliha @bartekn this is ready for review again

File: services/horizon/internal/db2/history/transaction.go [hunk 9]
```diff
@@ -26,6 +28,36 @@ func (q *Q) TransactionByHash(dest interface{}, hash string) error {
...
+	}
+
+	in := fmt.Sprintf("ht.id IN (%s)", sq.Placeholders(len(ids)))
+
+	whereArgs := make([]interface{}, len(ids))
+	for i, id := range ids {
+		whereArgs[i] = id
+	}
+
+	sql := selectTransaction.Where(in, whereArgs...)
```
Comment (bartekn) [inline services/horizon/internal/db2/history/transaction.go]: After approving I realized (looking at a not related code) that this can be simplified to:
```go
	sql := selectTransaction.Where(map[string]interface{}{
		"ht.id": ids,
	})
```
